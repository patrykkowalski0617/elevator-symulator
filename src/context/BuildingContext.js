import React, { createContext, useState } from "react";

export const BuildingContext = createContext();

const BuildingContextProvider = props => {
    const [numberOfFloors, setNumberOfFloors] = useState(8);

    return (
        <BuildingContext.Provider
            value={{
                numberOfFloors,
                setNumberOfFloors
            }}
        >
            {props.children}
        </BuildingContext.Provider>
    );
};

export default BuildingContextProvider;
