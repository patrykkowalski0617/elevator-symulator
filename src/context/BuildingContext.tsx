import React, { createContext, useState } from "react";

interface IContextProps {
    numberOfFloors: number;
    setNumberOfFloors: (n: number) => void;
    numberOfCars: number;
    setNnmberOfCars: (n: number) => void;
    automotion: boolean;
    setAutomotion: (b: boolean) => void;
    creatingStickMan: boolean;
    setCreatingStickMan: (arr: boolean) => void;
}

export const BuildingContext = createContext({} as IContextProps);

const BuildingContextProvider = (props: { children: React.ReactNode }) => {
    const [numberOfFloors, setNumberOfFloors] = useState(11);
    const howManyCars = Math.floor(numberOfFloors / 5);
    const [numberOfCars, setNnmberOfCars] = useState(howManyCars);
    const [automotion, setAutomotion] = useState(false);
    const [creatingStickMan, setCreatingStickMan] = useState<boolean>(false);

    return (
        <BuildingContext.Provider
            value={{
                numberOfFloors,
                setNumberOfFloors,
                numberOfCars,
                setNnmberOfCars,
                automotion,
                setAutomotion,
                creatingStickMan,
                setCreatingStickMan
            }}
        >
            {props.children}
        </BuildingContext.Provider>
    );
};

export default BuildingContextProvider;
