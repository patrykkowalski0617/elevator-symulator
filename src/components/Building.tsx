import React, { useContext } from "react";
import { BuildingContext } from "../context/BuildingContext";
import Floors from "./Floors";
import styled from "styled-components";
import Shaft from "./Shaft";
import StickManForm from "./StickManForm";

const BuildingStyled = styled.div`
    height: 90vh;
    width: 70vw;
    max-width: 600px;
    min-width: 320px;
    margin: 5vh auto;
    background: #000;
    position: relative;
    display: flex;
    flex-flow: row wrap;
    border-style: solid;
    border-width: 6px 4px;
`;

const Building: React.FC = () => {
    const { numberOfFloors, numberOfCars, creatingStickMan } = useContext(
        BuildingContext
    );

    // Sizes
    const carWidth: number = 10;
    const floorWidth: number = (100 - carWidth * numberOfCars) / 2;

    return (
        <>
            {typeof creatingStickMan === "number" ? (
                <StickManForm floorNumber={creatingStickMan} />
            ) : null}
            <BuildingStyled>
                <Floors role={"enter-floor"} floorWidth={floorWidth}></Floors>
                <Shaft
                    numberOfCars={numberOfCars}
                    carWidth={carWidth}
                    numberOfFloors={numberOfFloors}
                ></Shaft>
                <Floors role={"exit-floor"} floorWidth={floorWidth}></Floors>
            </BuildingStyled>
        </>
    );
};

export default Building;
