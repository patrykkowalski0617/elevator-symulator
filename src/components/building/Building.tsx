import React, { useContext } from "react";
import { BuildingContext } from "../../context/BuildingContext";
import FloorsStack from "../floors_stack/FloorsStack";
import { BuildingStyled } from "./BuildingStyled";
import Shaft from "../shaft/Shaft";
import StickManForm from "../form_stickman/FormStickMan";

/**
 * TO DO
 * @Sizes to be removed and used from BuildingContext
 */
const Building: React.FC = () => {
    const { numberOfFloors, numberOfCars, creatingStickMan } = useContext(
        BuildingContext
    );

    // Sizes
    const carWidth: string = "64px";
    const floorWidth: string = ` calc((100% - ${carWidth} * ${numberOfCars}) / 2)`;

    return (
        <>
            {typeof creatingStickMan === "number" ? (
                <StickManForm floorNumber={creatingStickMan} />
            ) : null}
            <BuildingStyled>
                <FloorsStack
                    role={"enter-floor"}
                    floorWidth={floorWidth}
                ></FloorsStack>
                <Shaft
                    numberOfCars={numberOfCars}
                    carWidth={carWidth}
                    numberOfFloors={numberOfFloors}
                ></Shaft>
                <FloorsStack
                    role={"exit-floor"}
                    floorWidth={floorWidth}
                ></FloorsStack>
            </BuildingStyled>
        </>
    );
};

export default Building;
