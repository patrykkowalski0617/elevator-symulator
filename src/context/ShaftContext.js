import React, { createContext, useState, useContext } from "react";
import { BuildingContext } from "./BuildingContext";

export const ShaftContext = createContext();

const ShaftContextProvider = props => {
    const { numberOfCars } = useContext(BuildingContext);
    const initArr = val => Array(numberOfCars).fill(val);

    const [carCurrenFloorCon, setCarCurrenFloorCon] = useState(initArr(0)); // next floor on way or floor that has been just reached, but not floor which has been just left

    // console.log(carCurrenFloorCon);

    return (
        <ShaftContext.Provider
            value={{
                carCurrenFloorCon,
                setCarCurrenFloorCon
            }}
        >
            {props.children}
        </ShaftContext.Provider>
    );
};

export default ShaftContextProvider;
