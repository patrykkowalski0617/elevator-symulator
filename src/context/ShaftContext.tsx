import React, { createContext, useState, useContext, useEffect } from "react";
import { BuildingContext } from "./BuildingContext";

interface IContextProps {
    allCarsCurrentFloor: number[];
    updateCarCurrentFloor: (carId: number, currentFloor: number) => void;
    allCarsFloorAssignments: number[][];
    addCarFloorAssignment: (carId: number, floorNumber: number) => void;
    removeCarFloorAssignment: (carId: number, floorNumber: number) => void;
    allCarsState: string[];
    updateCarState: (carId: number, state: string) => void;
    allCarsStickMans: {
        lifeState: string;
        destination: number;
        carId: number | null;
        placeInCar: number | null;
    }[][];
    addPassengers: (
        data: {
            destination: number;
            direction: string | null;
            lifeState: string;
            carId: number | null;
            placeInCar: number | null;
        }[]
    ) => void;
    allCarsDirection: string | null[];
    updateCarDirection: (carId: number, direction: string) => void;
}

export const ShaftContext = createContext({} as IContextProps);

const ShaftContextProvider = (props: { children: React.ReactNode }) => {
    const { numberOfCars } = useContext(BuildingContext);
    const initArr = (val: any) => Array(numberOfCars).fill(val);

    // next floor on way or floor that has been just reached, but not floor which has been just left
    const [allCarsCurrentFloor, setAllCarsCurrentFloor] = useState(initArr(0));
    // ready, go-up, go-down, go-up-door-open, go-down-door-open
    const [allCarsState, setAllCarsState] = useState(initArr("ready"));
    const [allCarsFloorAssignments, setAllCarsFloorAssignments] = useState(
        initArr([])
    );
    const [allCarsStickMans, setAllCarsStickMans] = useState(initArr([]));
    const [allCarsDirection, setAllCarsDirection] = useState(initArr(null));

    const updateCarCurrentFloor = (carId: number, currentFloor: number) => {
        const _allCarsCurrentFloor = [...allCarsCurrentFloor];
        _allCarsCurrentFloor.splice(carId, 1, currentFloor);
        setAllCarsCurrentFloor([..._allCarsCurrentFloor]);
    };

    const updateCarState = (carId: number, state: string) => {
        const _allCarsState = [...allCarsState];
        _allCarsState.splice(carId, 1, state);
        setAllCarsState([..._allCarsState]);
    };

    const addCarFloorAssignment = (carId: number, floorNumber: number) => {
        if (carId !== null) {
            const _allCarsFloorAssignments = [...allCarsFloorAssignments];
            const carFloorAssignment = [..._allCarsFloorAssignments[carId]];
            const carFloorAssignmentUpdated = [
                ...carFloorAssignment,
                floorNumber
            ];
            _allCarsFloorAssignments.splice(
                carId,
                1,
                carFloorAssignmentUpdated
            );
            setAllCarsFloorAssignments([..._allCarsFloorAssignments]);
        }
    };

    const [_allCarsStickMans, set_allCarsStickMans] = useState(
        allCarsStickMans
    );
    useEffect(() => {
        set_allCarsStickMans(allCarsStickMans);
    }, [allCarsStickMans]);
    const addPassengers = (
        data: {
            destination: number;
            direction: string | null;
            lifeState: string;
            carId: number | null;
            placeInCar: number | null;
        }[]
    ) => {
        if (data[0]) {
            const carId =
                typeof data[0].carId === "number" ? data[0].carId : -1;
            if (allCarsStickMans[carId].length < 4) {
                const carPassengers = [..._allCarsStickMans[carId], ...data];
                _allCarsStickMans.splice(carId, 1, carPassengers);
                setAllCarsStickMans([..._allCarsStickMans]);
            } else {
                console.warn("car is full!");
            }
        } else {
            console.warn(
                "there are no stickmen on this floor who have the same direction as the car!"
            );
        }
    };

    const removeCarFloorAssignment = (carId: number, floorNumber: number) => {
        const _allCarsFloorAssignments = [...allCarsFloorAssignments];
        const carFloorAssignment = [..._allCarsFloorAssignments[carId]];
        const index = carFloorAssignment.indexOf(floorNumber);
        carFloorAssignment.splice(index, 1);
        const carFloorAssignmentUpdated = [...carFloorAssignment];
        _allCarsFloorAssignments.splice(carId, 1, carFloorAssignmentUpdated);
        setAllCarsFloorAssignments([..._allCarsFloorAssignments]);
    };

    const updateCarDirection = (carId: number, direction: string) => {
        const _allCarsDirection = allCarsDirection;
        _allCarsDirection.splice(carId, 1, direction);
        setAllCarsDirection([..._allCarsDirection]);
    };

    return (
        <ShaftContext.Provider
            value={{
                allCarsCurrentFloor,
                updateCarCurrentFloor,
                allCarsFloorAssignments,
                addCarFloorAssignment,
                removeCarFloorAssignment,
                allCarsState,
                updateCarState,
                allCarsStickMans,
                addPassengers,
                allCarsDirection,
                updateCarDirection
            }}
        >
            {props.children}
        </ShaftContext.Provider>
    );
};

export default ShaftContextProvider;
