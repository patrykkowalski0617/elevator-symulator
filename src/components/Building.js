import React, { useRef, useContext } from "react";
import { BuildingContext } from "../context/BuildingContext";
import Floors from "./Floors";
import styled from "styled-components";
import Shaft from "./Shaft";
import { ShaftContext } from "../context/ShaftContext";

const BuildingStyled = styled.div`
    height: 90vh;
    width: 70vw;
    overflow: hidden;
    max-width: 600px;
    min-width: 320px;
    margin: 5vh auto;
    background: #000;
    position: relative;
    display: flex;
    flex-flow: row wrap;
    border-style: solid;
    border-width: 6px 4px;
`;

// const Roof = styled.div`
//     padding-left: 50%;
//     padding-bottom: 10%;
//     position: absolute;
//     right: 0;
//     left: 0;
//     z-index: -1000;
//     &:after {
//         content: "${props => props.text}";
//         display: block;
//         width: calc(100% - 25%);
//         height: 0;
//         margin-left: calc(-50% - 25% + 25px);
//         border-left: 25% solid transparent;
//         border-right: 25% solid transparent;
//         border-bottom: 25% solid #992222;
//         position: absolute;
//         bottom: 100%;
//     }
// `;

export default function Building() {
    const {
        numberOfFloors,
        numberOfCars,
        automotion,
        setAutomotion
    } = useContext(BuildingContext);

    const buildingRef = useRef();

    // Sizes
    const carWidth = 10;
    const floorWidth = (100 - carWidth * numberOfCars) / 2;

    const { setAllCarsCurrentFloor, allCarsState } = useContext(ShaftContext);

    return (
        <>
            <BuildingStyled ref={buildingRef}>
                {/* <Roof></Roof> */}
                <Floors role={"enter-floor"} floorWidth={floorWidth}></Floors>
                <Shaft
                    numberOfCars={numberOfCars}
                    carWidth={carWidth}
                    numberOfFloors={numberOfFloors}
                ></Shaft>
                <Floors role={"exit-floor"} floorWidth={floorWidth}></Floors>
            </BuildingStyled>
            {/* <div
                style={{
                    textAlign: "center",
                    position: "fixed",
                    bottom: "0",
                    width: "70%",
                    display: "flex",
                    flexWrap: "wrap",
                    transform: "translateX(-50%)",
                    margin: "0 50%"
                }}
            >
                <button
                    onClick={() => {
                        setAllCarsCurrentFloor([0, 0]);
                    }}
                    style={{ width: "25%", height: "30px" }}
                >
                    BOTTOM
                </button>
                <button
                    onClick={() => {
                        setAllCarsCurrentFloor([4, 4]);
                    }}
                    style={{ width: "25%", height: "30px" }}
                >
                    MID
                </button>
                <button
                    onClick={() => {
                        setAllCarsCurrentFloor([7, 7]);
                    }}
                    style={{ width: "25%", height: "30px" }}
                >
                    TOP
                </button>
                <button
                    onClick={() => {
                        setAllCarsCurrentFloor([7, 0]);
                    }}
                    style={{ width: "25%", height: "30px" }}
                >
                    TOP / BOTTOM
                </button>
                <button
                    onClick={() => {
                        setAllCarsCurrentFloor([0, 4]);
                    }}
                    style={{ width: "25%", height: "30px" }}
                >
                    BOTTOM / MID
                </button>
                <button
                    onClick={() => {
                        setAutomotion(!automotion);
                    }}
                    style={{ width: "25%", height: "30px" }}
                >
                    AUTOMOTION
                </button>

                <button
                    onClick={() => {
                        alert(allCarsState);
                    }}
                    style={{ width: "25%", height: "30px" }}
                >
                    PAUSE
                </button>
            </div> */}
        </>
    );
}
