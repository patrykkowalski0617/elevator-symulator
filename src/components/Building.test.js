import React from "react";
import Building from "./Building";
import renderer from "react-test-renderer";
import ShaftContext from "../context/ShaftContext";
import BuildingContext from "../context/BuildingContext";

test("Building - render and match snapshot", () => {
    const tree = renderer.create(
        <BuildingContext>
            <ShaftContext>
                <Building></Building>
            </ShaftContext>
        </BuildingContext>
    );
    expect(tree.toJSON()).toMatchSnapshot();
});
