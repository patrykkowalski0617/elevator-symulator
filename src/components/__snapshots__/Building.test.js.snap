// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Building - render and match snapshot 1`] = `
<div
  className="sc-Axmtr bVVBwG"
>
  <div
    className="sc-AxiKw kyKEfW"
    role="enter-floor"
  >
    <div
      className="sc-AxjAm cEScls"
    >
      <button
        data-floor-number={9}
        onClick={[Function]}
        style={null}
      >
        Call car
      </button>
      <button
        data-floor-number={9}
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <button
        data-floor-number={8}
        onClick={[Function]}
        style={null}
      >
        Call car
      </button>
      <button
        data-floor-number={8}
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <button
        data-floor-number={7}
        onClick={[Function]}
        style={null}
      >
        Call car
      </button>
      <button
        data-floor-number={7}
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <button
        data-floor-number={6}
        onClick={[Function]}
        style={null}
      >
        Call car
      </button>
      <button
        data-floor-number={6}
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <button
        data-floor-number={5}
        onClick={[Function]}
        style={null}
      >
        Call car
      </button>
      <button
        data-floor-number={5}
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <button
        data-floor-number={4}
        onClick={[Function]}
        style={null}
      >
        Call car
      </button>
      <button
        data-floor-number={4}
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <button
        data-floor-number={3}
        onClick={[Function]}
        style={null}
      >
        Call car
      </button>
      <button
        data-floor-number={3}
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <button
        data-floor-number={2}
        onClick={[Function]}
        style={null}
      >
        Call car
      </button>
      <button
        data-floor-number={2}
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <button
        data-floor-number={1}
        onClick={[Function]}
        style={null}
      >
        Call car
      </button>
      <button
        data-floor-number={1}
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <button
        data-floor-number={0}
        onClick={[Function]}
        style={null}
      >
        Call car
      </button>
      <button
        data-floor-number={0}
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
  </div>
  <div
    className="sc-AxheI kVJGRi"
  >
    <div
      className="sc-AxhCb bXgSMk"
      style={
        Object {
          "transform": "translateY(calc(-0%))",
        }
      }
    >
      <button
        onClick={[Function]}
      >
        start
      </button>
      <div
        className="sc-AxhUy ZMmr"
        speed={0}
      >
        <div
          className="sc-AxgMl jwFsLE"
          speed={0}
        />
      </div>
      <p>
        T: 
        t
      </p>
      <p>
        Now: 
        -
      </p>
    </div>
  </div>
  <div
    className="sc-AxheI kVJGRi"
  >
    <div
      className="sc-AxhCb bXgSMk"
      style={
        Object {
          "transform": "translateY(calc(-0%))",
        }
      }
    >
      <button
        onClick={[Function]}
      >
        start
      </button>
      <div
        className="sc-AxhUy ZMmr"
        speed={0}
      >
        <div
          className="sc-AxgMl jwFsLE"
          speed={0}
        />
      </div>
      <p>
        T: 
        t
      </p>
      <p>
        Now: 
        -
      </p>
    </div>
  </div>
  <div
    className="sc-AxiKw kyKEfW"
    role="exit-floor"
  >
    <div
      className="sc-AxjAm cEScls"
    >
      <p>
        9
      </p>
      <input
        style={
          Object {
            "height": "15px",
            "width": "15px",
          }
        }
        type="checkbox"
      />
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <p>
        8
      </p>
      <input
        style={
          Object {
            "height": "15px",
            "width": "15px",
          }
        }
        type="checkbox"
      />
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <p>
        7
      </p>
      <input
        style={
          Object {
            "height": "15px",
            "width": "15px",
          }
        }
        type="checkbox"
      />
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <p>
        6
      </p>
      <input
        style={
          Object {
            "height": "15px",
            "width": "15px",
          }
        }
        type="checkbox"
      />
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <p>
        5
      </p>
      <input
        style={
          Object {
            "height": "15px",
            "width": "15px",
          }
        }
        type="checkbox"
      />
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <p>
        4
      </p>
      <input
        style={
          Object {
            "height": "15px",
            "width": "15px",
          }
        }
        type="checkbox"
      />
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <p>
        3
      </p>
      <input
        style={
          Object {
            "height": "15px",
            "width": "15px",
          }
        }
        type="checkbox"
      />
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <p>
        2
      </p>
      <input
        style={
          Object {
            "height": "15px",
            "width": "15px",
          }
        }
        type="checkbox"
      />
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <p>
        1
      </p>
      <input
        style={
          Object {
            "height": "15px",
            "width": "15px",
          }
        }
        type="checkbox"
      />
    </div>
    <div
      className="sc-AxjAm cEScls"
    >
      <p>
        0
      </p>
      <input
        style={
          Object {
            "height": "15px",
            "width": "15px",
          }
        }
        type="checkbox"
      />
    </div>
  </div>
</div>
`;
